handler/deployment/deployment.go

```diff
@@ -224,6 +249,47 @@ func (d *Deployment) deploymentSuccess(ctx context.Context, correlationID, evalu
 	}
 }
 
+func (d *Deployment) systemDeploymentSuccess(ctx context.Context, correlationID, evaluationID, jobID string, jobVersion uint64) error {
+	ticker := time.Tick(time.Second * 1)
+	timeout := time.After(d.timeout)
+	minLogData := log.Data{"evaluation": evaluationID, "job": jobID, "job_version": jobVersion}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Event(ctx, "bailing on deployment status", log.ERROR, minLogData)
+			return &AbortedError{EvaluationID: evaluationID, CorrelationID: correlationID}
+		case <-timeout:
+			return &TimeoutError{Action: "deployment"}
+		case <-ticker:
+			var allocations []api.AllocationListStub
+			if err := d.get(fmt.Sprintf(allocationsURL, d.endpoint, jobID), &allocations); err != nil {
+				return err
+			}
+
+			if len(allocations) == 0 {
+				log.Event(ctx, "deployment failed - no allocations", log.ERROR, minLogData)
+				return &AbortedError{EvaluationID: evaluationID, CorrelationID: correlationID}
+			}
+
+			updateComplete := true
+			for _, allocation := range allocations {
+				if allocation.JobVersion != jobVersion || allocation.ClientStatus != structs.AllocClientStatusRunning {
```

Comment:
I didn't really like assuming the `updateComplete` as `true` either. But the aim of the check is to make sure that all allocations are the current version and are 'running'.
Defaulting to `false` and setting to `true` as you've described the check would pass as long as one allocation was 'running' the new version, rather than requiring all of them.
Perhaps we start with the assumption that 0 of the allocations are running the new version, then count the ones that are, and check the count against the total number of allocations at the end. Something like:
```suggestion
				updatedAllocations := 0
				...
				if allocation.JobVersion == jobVersion && allocation.ClientStatus == structs.AllocClientStatusRunning {
				       updatedAllocations += 1
				...
				len(allocations) == updatedAllocations
```
allows for the initial assumption that the deployment hasn't completed and only passes when all allocations are updated?

Reply:
Gotcha. Yes, a count sounds better.

Reply:
Ok, I've made the change to use a count instead.